File: backend/src/threatlens/api/routes/workspace.py
"""Investigation Workspace routes: save, load, update, delete, and list completed investigations.

A workflow and persistence layer over the existing analytical pipeline — not a
new intelligence engine. Every endpoint here is thin transport over
:class:`~threatlens.workspace.service.WorkspaceService`; the workspace never
computes an investigation result, it only stores and retrieves whatever the
caller already produced via ``/investigate``, ``/detections``, or (in a future
phase) correlation. No authentication — single-user, self-hosted.
"""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query

from ...entities.types import EntityType
from ...graph import EvidenceGraph, GraphService
from ...reasoning import Severity
from ...timeline import Timeline, TimelineService
from ...workspace import (
    InvestigationNotFoundError,
    LocalFileStorage,
    SaveInvestigationRequest,
    UpdateInvestigationRequest,
    WorkspaceInvestigation,
    WorkspaceService,
    WorkspaceSettings,
    WorkspaceStatus,
)
from ..schemas import WorkspaceListItem, WorkspaceListResponse

router = APIRouter()

# Process-wide workspace service, backed by local-file storage. Built once; the
# storage root is created on first use (see LocalFileStorage.__init__).
_workspace_settings = WorkspaceSettings.from_env()
_workspace_service = WorkspaceService(LocalFileStorage(_workspace_settings.storage_dir))

# The Timeline Framework (Phase 8.1) is stateless — a pure function of
# whatever WorkspaceInvestigation it's given — so one shared instance is
# just a namespace, not a cache; still built once for consistency with every
# other *_service singleton in this module.
_timeline_service = TimelineService()

# The Evidence Relationship Graph Framework (Phase 8.2) is likewise a
# stateless pure function of one WorkspaceInvestigation — a sibling of the
# Timeline service, not a consumer of it.
_graph_service = GraphService()


def get_workspace_service() -> WorkspaceService:
    """Provide the Workspace service (overridable in tests)."""
    return _workspace_service


def get_timeline_service() -> TimelineService:
    """Provide the Timeline service (overridable in tests)."""
    return _timeline_service


def get_graph_service() -> GraphService:
    """Provide the Graph service (overridable in tests)."""
    return _graph_service


def _to_list_item(record: WorkspaceInvestigation) -> WorkspaceListItem:
    return WorkspaceListItem(
        id=record.id,
        title=record.title,
        created_at=record.created_at,
        updated_at=record.updated_at,
        status=record.status,
        tags=record.tags,
        summary=record.summary,
        severity=record.severity,
        investigation_type=record.investigation_type,
    )


@router.post(
    "/api/v1/workspace",
    response_model=WorkspaceInvestigation,
    status_code=201,
)
def save_investigation(
    request: SaveInvestigationRequest,
    service: Annotated[WorkspaceService, Depends(get_workspace_service)],
) -> WorkspaceInvestigation:
    """Save a completed investigation. ``investigation_summary``/``detection_package``/
    ``correlation_summary`` are attached verbatim — nothing is recomputed."""
    return service.save(request)


@router.get("/api/v1/workspace", response_model=WorkspaceListResponse)
def list_investigations(
    service: Annotated[WorkspaceService, Depends(get_workspace_service)],
    status: Annotated[WorkspaceStatus | None, Query()] = None,
    severity: Annotated[Severity | None, Query()] = None,
    investigation_type: Annotated[EntityType | None, Query()] = None,
    tag: Annotated[str | None, Query()] = None,
    q: Annotated[
        str | None, Query(description="Case-insensitive search over title/summary/tags")
    ] = None,
) -> WorkspaceListResponse:
    """List saved investigations (metadata only), most recently updated first.

    All filters are optional and combine with AND. Every match is a pure,
    in-memory filter over already-persisted metadata.
    """
    records = service.list(
        status=status,
        severity=severity,
        investigation_type=investigation_type,
        tag=tag,
        query=q,
    )
    items = [_to_list_item(r) for r in records]
    return WorkspaceListResponse(investigations=items, total=len(items))


@router.get("/api/v1/workspace/{investigation_id}", response_model=WorkspaceInvestigation)
def get_investigation(
    investigation_id: UUID,
    service: Annotated[WorkspaceService, Depends(get_workspace_service)],
) -> WorkspaceInvestigation:
    """Load one saved investigation, including every attached engine output."""
    try:
        return service.get(investigation_id)
    except InvestigationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/api/v1/workspace/{investigation_id}/timeline", response_model=Timeline)
def get_investigation_timeline(
    investigation_id: UUID,
    workspace: Annotated[WorkspaceService, Depends(get_workspace_service)],
    timeline: Annotated[TimelineService, Depends(get_timeline_service)],
) -> Timeline:
    """Derive the read-only timeline for one saved investigation.

    Consumes the saved record's existing ``investigation_summary`` verbatim;
    never mutates it and never persists timeline data of its own — a
    repeated call against an unchanged saved record always returns a
    byte-identical ``Timeline``. ``404`` if the investigation doesn't exist.
    """
    try:
        record = workspace.get(investigation_id)
    except InvestigationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return timeline.build(record)


@router.get("/api/v1/workspace/{investigation_id}/graph", response_model=EvidenceGraph)
def get_investigation_graph(
    investigation_id: UUID,
    workspace: Annotated[WorkspaceService, Depends(get_workspace_service)],
    graph: Annotated[GraphService, Depends(get_graph_service)],
) -> EvidenceGraph:
    """Derive the read-only evidence graph for one saved investigation.

    Consumes the saved record's existing ``investigation_summary`` and
    ``correlation_summary`` verbatim; never mutates either and never
    persists graph data of its own — a repeated call against an unchanged
    saved record always returns a byte-identical ``EvidenceGraph``. ``404``
    if the investigation doesn't exist.
    """
    try:
        record = workspace.get(investigation_id)
    except InvestigationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return graph.build(record)


@router.put("/api/v1/workspace/{investigation_id}", response_model=WorkspaceInvestigation)
def update_investigation(
    investigation_id: UUID,
    request: UpdateInvestigationRequest,
    service: Annotated[WorkspaceService, Depends(get_workspace_service)],
) -> WorkspaceInvestigation:
    """Partially update a saved investigation's metadata (and/or re-attach an output).

    Only fields present in the request body change; an omitted field keeps its
    current value (see :class:`~threatlens.workspace.models.UpdateInvestigationRequest`).
    """
    try:
        return service.update(investigation_id, request)
    except InvestigationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.delete("/api/v1/workspace/{investigation_id}", status_code=204)
def delete_investigation(
    investigation_id: UUID,
    service: Annotated[WorkspaceService, Depends(get_workspace_service)],
) -> None:
    """Delete a saved investigation."""
    try:
        service.delete(investigation_id)
    except InvestigationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

File: frontend/lib/api/workspace.ts
// Investigation Workspace (Phase 8.0) — a persistence layer over completed
// investigations. Not a new intelligence engine: this module only stores and
// retrieves what /investigate, /detections, and (in a future phase)
// /correlation already produced.

import { del, get, post, put } from "./client";
import type { DetectionPackage } from "./detection";
import type { EntityType, InvestigationSummary } from "./investigation";

export type WorkspaceStatus = "open" | "in_progress" | "closed" | "archived";

/**
 * Mirrors `threatlens.correlation.models.CorrelationSummary` by name only. No
 * UI renders its fields yet — Correlation isn't wired into `/investigate`, so
 * this is almost always absent on a saved investigation. Kept deliberately
 * opaque until a future phase gives it a real consumer and a precise shape.
 */
export type CorrelationSummary = Record<string, unknown>;

export interface WorkspaceInvestigation {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
  status: WorkspaceStatus;
  tags: string[];
  summary: string | null;
  severity: number | null;
  investigation_type: EntityType;
  investigation_summary: InvestigationSummary | null;
  detection_package: DetectionPackage | null;
  correlation_summary: CorrelationSummary | null;
}

/** One row of `GET /workspace` — metadata only, no nested engine outputs. */
export interface WorkspaceListItem {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
  status: WorkspaceStatus;
  tags: string[];
  summary: string | null;
  severity: number | null;
  investigation_type: EntityType;
}

export interface WorkspaceListResponse {
  investigations: WorkspaceListItem[];
  total: number;
}

export interface SaveInvestigationRequest {
  title: string;
  status?: WorkspaceStatus;
  tags?: string[];
  summary?: string | null;
  severity?: number | null;
  investigation_type: EntityType;
  investigation_summary?: InvestigationSummary | null;
  detection_package?: DetectionPackage | null;
  correlation_summary?: CorrelationSummary | null;
}

/** Every field is optional; an omitted field leaves the saved value unchanged. */
export interface UpdateInvestigationRequest {
  title?: string;
  status?: WorkspaceStatus;
  tags?: string[];
  summary?: string | null;
  severity?: number | null;
  investigation_type?: EntityType;
  investigation_summary?: InvestigationSummary | null;
  detection_package?: DetectionPackage | null;
  correlation_summary?: CorrelationSummary | null;
}

export interface WorkspaceListFilters {
  status?: WorkspaceStatus;
  severity?: number;
  investigation_type?: EntityType;
  tag?: string;
  q?: string;
}

function buildQuery(filters: WorkspaceListFilters): string {
  const params = new URLSearchParams();
  if (filters.status) params.set("status", filters.status);
  if (filters.severity !== undefined) params.set("severity", String(filters.severity));
  if (filters.investigation_type) params.set("investigation_type", filters.investigation_type);
  if (filters.tag) params.set("tag", filters.tag);
  if (filters.q) params.set("q", filters.q);
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

/** Save a completed investigation. Attaches whatever engine outputs the caller already has. */
export function saveInvestigation(
  request: SaveInvestigationRequest,
  signal?: AbortSignal,
): Promise<WorkspaceInvestigation> {
  return post<WorkspaceInvestigation>("/workspace", request, signal);
}

/** List saved investigations (metadata only), optionally filtered; most recently updated first. */
export function listInvestigations(
  filters: WorkspaceListFilters = {},
  signal?: AbortSignal,
): Promise<WorkspaceListResponse> {
  return get<WorkspaceListResponse>(`/workspace${buildQuery(filters)}`, signal);
}

/** Load one saved investigation, including every attached engine output. */
export function getInvestigation(
  id: string,
  signal?: AbortSignal,
): Promise<WorkspaceInvestigation> {
  return get<WorkspaceInvestigation>(`/workspace/${encodeURIComponent(id)}`, signal);
}

/** Partially update a saved investigation's metadata (and/or re-attach an output). */
export function updateInvestigation(
  id: string,
  request: UpdateInvestigationRequest,
  signal?: AbortSignal,
): Promise<WorkspaceInvestigation> {
  return put<WorkspaceInvestigation>(`/workspace/${encodeURIComponent(id)}`, request, signal);
}

/** Delete a saved investigation. */
export function deleteInvestigation(id: string, signal?: AbortSignal): Promise<void> {
  return del(`/workspace/${encodeURIComponent(id)}`, signal);
}

// --- Investigation Timeline (Phase 8.1) ---
//
// A read-only, derived view over a saved investigation's existing evidence —
// never a new intelligence engine. An event exists only when the evidence it
// came from already carried an explicit timestamp; the UI renders exactly
// what the backend returns and never re-sorts, re-derives, or estimates a
// missing time itself.

/**
 * `event_type`/`source_type` are left as plain `string` rather than exact
 * unions — mirrors `Evidence.type` in `./investigation`, which does the same
 * for the same reason: nothing in the UI branches on a specific value.
 */
export interface TimelineEvent {
  event_id: string;
  timestamp: string;
  event_type: string;
  title: string;
  description: string;
  source_type: string;
  source_id: string;
  severity: number | null;
  evidence_references: string[];
}

export interface Timeline {
  investigation_id: string;
  entity_type: EntityType;
  entity_value: string;
  generated_at: string;
  events: TimelineEvent[];
}

/** Derive the read-only timeline for one saved investigation. */
export function getInvestigationTimeline(
  id: string,
  signal?: AbortSignal,
): Promise<Timeline> {
  return get<Timeline>(`/workspace/${encodeURIComponent(id)}/timeline`, signal);
}

// --- Evidence Relationship Graph (Phase 8.2) ---
//
// A read-only, derived view over a saved investigation's existing evidence
// and correlation output — never a new intelligence engine. A node or edge
// exists only when the backend found supporting evidence or an explicit
// relationship; the UI renders exactly what it returns and never infers,
// merges, or lays out a connection itself.

/**
 * `node_type`/`relationship_type` are left as plain `string` rather than
 * exact unions — mirrors `TimelineEvent.event_type` above for the same
 * reason: the underlying vocabularies genuinely differ by source, and
 * nothing in the UI branches on a specific value.
 */
export interface GraphNode {
  node_id: string;
  node_type: string;
  label: string;
  value: string;
  severity: number | null;
  source_references: string[];
  metadata: Record<string, unknown>;
}

export interface GraphEdge {
  edge_id: string;
  source_node_id: string;
  target_node_id: string;
  relationship_type: string;
  explanation: string;
  evidence_references: string[];
  source_references: string[];
}

export interface EvidenceGraph {
  investigation_id: string;
  entity_type: EntityType;
  entity_value: string;
  generated_at: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
  node_count: number;
  edge_count: number;
  graph_version: string;
}

/** Derive the read-only evidence relationship graph for one saved investigation. */
export function getInvestigationGraph(
  id: string,
  signal?: AbortSignal,
): Promise<EvidenceGraph> {
  return get<EvidenceGraph>(`/workspace/${encodeURIComponent(id)}/graph`, signal);
}
